File: containerWithMaxWater.java
//time: O(N) where N is the no of elements
//space: O(1)
//approach: Two pointers

class Solution {
    public int maxArea(int[] nums) {
        int left = 0;
        int n = nums.length;
        int right = n -1;
        int area = 0;
        while(left < right) {
            area = Math.max(area, Math.min(nums[left], nums[right]) * Math.abs((right - left)));
            if(nums[left] < nums[right]) {
                left++;
            } else if(nums[left] > nums[right]) {
                right--;
            } else {
                left++; right--;
            }
        }
        return area;
    }
}

File: ContainerWithMaxWater.py
# TC=O(n)
# SC = O(1)
class Solution:
    def maxArea(self, height: List[int]) -> int:
        l = 0
        r = len(height) - 1
        res = 0
        while l < r:
            length = min(height[l], height[r])
            area = length * (r - l)
            res = max(res, area)
            if height[l] < height[r]:
                l += 1
            else:
                r -= 1
        return res
